src/monitor.ts

```diff
@@ -0,0 +1,425 @@
+ * network node.
+ */
+export interface IMonitoredNodeStatus {
+  address: string;
+  alive: boolean;
+  lastUpdate: Date;
+  lastResponseId: number;
+  failureCount: number;
+  networkId?: string;
+  reviveCmd?: string;
+  reviveArgs?: string[];
+}
+
+/**
+ * An internal interface representing the response of the ping command to the
+ * network node.
+ */
+interface INetVersionResponse {
+  id: number;
+  result: string;
+}
+
+/**
+ * The actual monitor implementation. It contains an express application and
+ * last known node status.
+ */
+export class Monitor {
+  /**
+   * The interval at which the monitor pings the instances. If the last update
+   * time was longer than this interval, the instance is considered dead.
+   */
+  private heartbeatInterval: number;
+
+  /**
+   * Number of failures the monitor should tolerate before considering a node as
+   * dead.
+   */
+  private failureTolerance: number;
+
+  // Path to the configuration file.
+  private configPath: string;
+
+  // The express app
+  private app: express.Application;
+
+  // The http server
+  private server?: http.Server;
+
+  // The current state of the nodes
+  private nodesStatus: IMonitoredNodeStatus[];
+
+  // The timer used to ping nodes
+  private timer?: NodeJS.Timer;
+
+  /**
+   * Construct a monitor for the given nodes, open an HTTP server on the given
+   * port for incoming requests.
+   * @param nodes The configurations for each monitored node
+   * @param configPath The path to the configuration file
+   * @param heartbeatInterval Number of milliseconds between each ping request
+   * @param failureTolerance Number of failures monitor should tolerate before
+   * considering a node dead, and trying to revive it
+   */
+  constructor(nodes: IMonitoredNodeConfig[],
+              configPath: string,
+              heartbeatInterval?: number,
+              failureTolerance?: number) {
+    this.app = express();
+    this.setupRouting();
+    this.server = undefined;
+    this.timer = undefined;
+
+    this.nodesStatus = [];
+
+    // Why do we pass both config data and config path in? Because the
+    // constructor cannot be asynchronous... And since parsing the configuration
+    // can potentially block, we don't want to parse that data synchronously
+    // either. This is a compromise.
+    this.configPath = configPath;
+    for (const node of nodes) {
+      this.nodesStatus.push(this.configToInitialNodeStatus(node));
+    }
+
+    if (typeof heartbeatInterval !== 'undefined') {
+      this.heartbeatInterval = heartbeatInterval;
+    } else {
+      this.heartbeatInterval = 5000;
+    }
+
+    if (typeof failureTolerance !== 'undefined') {
+      this.failureTolerance = failureTolerance;
+    } else {
+      this.failureTolerance = 1;
+    }
+  }
+
+  /**
+   * Start the embedded HTTP server to process requests on the given port.
+   */
+  public start(port: number) {
+    this.timer = this.createPingTimer();
+    this.ping();
```

Comment:
do we need this ping too?

Reply:
Yeah because the monitor should have status immediately after startup. If we don't ping once here users will have to wait a few seconds.